### PR TITLE
Add FragmentCache functional

### PR DESCRIPTION
**Pull request by mj4444ru**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Fixed issues  | https://github.com/yiisoft/view/issues/10

#### DynamicContentAwareTrait.php
  - Fixed type mismatch error if evaluateDynamicContent method returns non-string result.

#### View.php
  - Minor fixes in PHPDOC.
  - Added private property FragmentCacheInterface $fragmentCache.
  - Added private property string $placeholderDynamicSignature. It contains a random signature to fix the signature intersection issue.
  - Added $fragmentCache initialization to the constructor.
  - The setPlaceholderSalt method is replaced by generatePlaceholderSignatures. This method populates the $placeholderSignature and $placeholderDynamicSignature properties.
  - Backslash added to basic php functions.
  - The renderDynamic method now uses a dynamic (random) signature, because when using a constant signature, an error occurred with the intersection of dynamic insert names when using data from the cache in another cached fragment.
  - Added $ instance parameter to the popDynamicContent method to check the correct cache closing sequence.
  - Added a public beginCache(string $id, array $params = [], array $vars = []): ?FragmentCache. This method starts fragment caching and returns an object of the FragmentCache class or null if the data is found in the cache. The parameters of the method are the same as those of the beginCache method of the FragmentCache class. If data is found in the cache, it will be automatically output to the stream. If the data is not found in the cache, it is necessary to generate new data and at the end call the endCache method on the returned object to save the data in the cache and finish caching.
  
#### WebView.php
  - Calls of methods registerMetaTag, registerLinkTag, registerCss, registerCssFile, registerJs, registerJsFile, registerJsVar, setCssFiles, setJsFiles, setTitle are now cached along with the main content.
  
#### Added Yiisoft\View\FragmentCacheInterface interface.
  - This interface is used in the Yiisoft\View\View class to implement caching.
  
#### Description of the implementation of the Yiisoft\View\FragmentCache class.
  - This class implements the Yiisoft\View\FragmentCacheInterface interface and is used in the Yiisoft\View\View class.
  - Public properties:
    - int $duration = 60 - Caching time (can be changed during content generation).
    - bool $enabled = true - Cache activity (can be changed during content generation to false, changing from false to true will have no effect).
  - Public methods
    - beginCache(?View $view, string $id, array $params = [], array $vars = []): self
      - The method returns a cloned instance of itself, tries to read the content from the cache, and if it fails, starts caching.
      - Method parameters
        - $view - a link to a Yiisoft\View\View instance to register itself as an object of the Yiisoft\View\DynamicContentAwareInterface.
        - $id - String unique identifier of the cached block.
        - $params - Array of parameters. Possible parameters:
          - duration (int) - Caching time.
          - enabled (bool) - Cache activity (ignored if caching was disabled earlier).
          - cache (Psr\SimpleCache\CacheInterface|Yiisoft\Cache\CacheInterface)
          - dependency - If you use a cache that supports the Yiisoft\Cache\CacheInterface interface, you can transfer dependencies here.
          - variations - Additional data defining the uniqueness (variability) of the data. Added to id for variability.
        - $vars - Array of variables used in the fragment. Variables are not saved in the cache.
      - endCache()
        - The method ends the previously started caching with the beginCache method.
      - getStatus(): int
        - The method returns the status of the cached object. Possible statuses:
          - STATUS_NO - This is the root object used to call the beginCache method.
          - STATUS_INIT - This is the cloned object during the initialization phase.
          - STATUS_BEGIN - Caching in progress (after calling beginCache).
          - STATUS_END - Caching completed (after calling endCache).
          - STATUS_IN_CACHE - The content was received from the cache, a call to endCache is required to output data to the stream (after calling beginCache).
      - renderVar(string $name): string
        - The method returns the value of the variable (the signature or the value itself).

#### Variable content
  - Sometimes it is necessary to cache a large fragment, but there is a line in it that may change. You can use the variations value of the beginCache method parameters, but unlike variations, using variables allows you to store one instance of cached data.
  - Variable content differs from dynamic content in that it is calculated not after the formation of all content, but before the start of caching and can interact with the current environment and data. Also, when re-caching, the top-level cache will cache the variable content as text, not as variable content.

## Configuration example:
```php
final class WebViewProvider extends ServiceProvider
{
    public function register(Container $container): void
    {
        $container->set(FragmentCacheInterface::class, FragmentCache::class);
        $container->set(WebView::class, function (ContainerInterface $container, Aliases $aliases, Theme $theme, EventDispatcherInterface $eventDispatcher, FragmentCacheInterface $fragmentCache, LoggerInterface $logger) {
            $webView = new WebView($aliases->get($this->basePath), $theme, $eventDispatcher, $fragmentCache, $logger);

            $defaultParameters = [];
            foreach ($this->defaultParameters as $key => $value) {
                $defaultParameters[$key] = $container->get($value);
            }
            $webView->setDefaultParameters($defaultParameters);

            return $webView;
        });
    }
}
```

## An example of using dynamic content:
```php
if ($cache = $this->beginCache('test/test')):

?> Data Data Data <?= $this->renderDynamic('return date("d.m.Y H:i");') ?> Data Data Data <?php

	$cache->endCache();
endif;
```

## An example of using variables (they are not saved in the cache):
```php
if ($cache = $this->beginCache('test/test', ['duration' => 3600], ['content' => $content])):

?> Data Data Data <?= $cache->renderVar('content') ?> Data Data Data <?php

	$cache->endCache();
endif;
```

## Tests and documentation
There are no tests and documentation.

### CODE REVIEW

File: src/View.php [hunk 1]
```diff
@@ -87,17 +93,17 @@ class View implements DynamicContentAwareInterface
...
 
     /**
-     * @var LoggerInterface $logger
+     * @var LoggerInterface
      */
     private LoggerInterface $logger;
 
     /**
-     * @var string $sourceLanguage
+     * @var string
```
Comment (roxblnfk) [inline src/View.php:106]: all comments that do not carry a meaningful description or payload should be removed

File: src/View.php [hunk 2]
```diff
@@ -76,6 +77,11 @@ class View implements DynamicContentAwareInterface
      */
     protected Theme $theme;
 
+    /**
+     * @var FragmentCache
```
Comment (roxblnfk) [inline src/View.php:81]: PhpDoc and prperty type are different. Is it ok?

Comment (mj4444ru) [reply src/View.php]: Thank. Fixed.

File: src/View.php (hunk 1 above)
Comment (mj4444ru) [reply src/View.php]: Thank. Fixed.

**Comment by samdark** [comment]
Thanks for contribution. There are 2 issues with this pull request:

1. It breaks current unit tests.
2. There are no tests for functionality added.

**Comment by mj4444ru** [comment]
> Thanks for contribution. There are 2 issues with this pull request:
> 
> 1. It breaks current unit tests.
> 2. There are no tests for functionality added.

Locally, Unit tests are all passed. What's wrong with them on GitHub?

**Comment by samdark** [comment]
Fixed tests. They pass locally now.

@terabytesoftw seems we have https://github.com/yiisoft/yii-widgets/blob/master/src/FragmentCache.php ... Would you please check it overall?

**Comment by terabytesoftw** [comment]
> Fixed tests. They pass locally now.
> 
> @terabytesoftw seems we have https://github.com/yiisoft/yii-widgets/blob/master/src/FragmentCache.php ... Would you please check it overall?

The idea of FragmentCache is to save a part of the view in Cache, this PR goes much further, it is practically a View Cache, which in the end can cause problems in the view with the record of assets and the operation of the view. The functionality could be updated in the FragmentCache Widget in yii-widgets.

**Comment by mj4444ru** [comment]
> > Fixed tests. They pass locally now.
> > @terabytesoftw seems we have https://github.com/yiisoft/yii-widgets/blob/master/src/FragmentCache.php ... Would you please check it overall?
> 
> The idea of `FragmentCache::class` is to save a part of the view in cache, this PR goes much further, it is practically a `view cache`, which in the end can cause problems in the view with the register of assets and the operation of the view. The functionality could be updated in the FragmentCache Widget in yii-widgets.

1) Explain what problems with assets registration are we talking about?
2) What are the benefits of implementing a widget (other than complication)?
3) The FragmentCache widget does not seem to work as expected. Most likely its current implementation will not be compatible with the current widget implementation.
4) The widget will definitely not work like this:
FragmentCache::begin(...);
FragmentCache::end();

**Comment by terabytesoftw** [comment]
> > > Fixed tests. They pass locally now.
> > > @terabytesoftw seems we have https://github.com/yiisoft/yii-widgets/blob/master/src/FragmentCache.php ... Would you please check it overall?
> > 
> > 
> > The idea of  is to save a part of the view in cache, this PR goes much further, it is practically a , which in the end can cause problems in the view with the register of assets and the operation of the view. The functionality could be updated in the FragmentCache Widget in yii-widgets.`FragmentCache::class``view cache`
> 
> 1. Explain what problems with assets registration are we talking about?
for example https://github.com/yiisoft/assets/issues/8 

> 2. What are the benefits of implementing a widget (other than complication)?
Having a simplified view, your code adds a lot of complexity, which can lead to errors in the operation of the view, 

> 3. The FragmentCache widget does not seem to work as expected. Most likely its current implementation will not be compatible with the current widget implementation.
The widget needs to be updated.

> 4. The widget will definitely not work like this:
>    FragmentCache::begin(...);
>    FragmentCache::end();
With the widget it is only used in the required view.

**Comment by mj4444ru** [comment]
> > > > Fixed tests. They pass locally now.
> > > > @terabytesoftw seems we have https://github.com/yiisoft/yii-widgets/blob/master/src/FragmentCache.php ... Would you please check it overall?
> > > 
> > > 
> > > The idea of  is to save a part of the view in cache, this PR goes much further, it is practically a , which in the end can cause problems in the view with the register of assets and the operation of the view. The functionality could be updated in the FragmentCache Widget in yii-widgets.`` FragmentCache::class``view cache ``
> 
> 1. Explain what problems with assets registration are we talking about?
>    for example [yiisoft/assets#8](https://github.com/yiisoft/assets/issues/8)
> 2. What are the benefits of implementing a widget (other than complication)?
>    Having a simplified view, your code adds a lot of complexity, which can lead to errors in the operation of the view,
> 3. The FragmentCache widget does not seem to work as expected. Most likely its current implementation will not be compatible with the current widget implementation.
>    The widget needs to be updated.
> 4. The widget will definitely not work like this:
>    FragmentCache::begin(...);
>    FragmentCache::end();
>    With the widget it is only used in the required view.
> 
> Note: It is a bad idea to cache all view values.

You misunderstand the meaning. This is not something that gets passed to the view. And the cache has nothing to do with the view.

>    for example [yiisoft/assets#8](https://github.com/yiisoft/assets/issues/8)
This is in no way related to the cache and takes place even without the cache.

>    Having a simplified view, your code adds a lot of complexity, which can lead to errors in the operation of the view,
Do you think there is more complex logic in the FragmentCache widget? If not, what's the point of discussing this?
You still cannot explain what you mean.

>    The widget needs to be updated.
It's easier to rewrite from scratch than to finish what is there.

>    With the widget it is only used in the required view.
This is very bad.

**Pull request closed without merging**